### PR TITLE
New 2020-03-01 API Version for Streaming Jobs

### PR DESCRIPTION
**Pull request by roslynlu**
Features added to this new swagger version:

	1. Dedicated cluster compatibility/linking streaming job 
	2. Dynamic scaling
	3. Job Compatibility level 1.2
	4. Service Bus output SystemProperties
	5. Parquet serialization type
	6. Time window and Size window output properties for blob output
	7. MSI for PBI
	8. MSI for Blob
	9. SQL DW output
	10. Edge support
	11. MSI for SQL DB
	12. Output data Lake MSI
	13. Input partition key
	14. Event Hub property columns
	15. SQL DB output max batch count and max writer count
	16. ContentStoragePolicy property + job storage account
	17. MSI for job storage account
	18. SQL reference input
	19. Input compression

<hr>

<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [x] new API version 
      - [ ] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  -  [x] Ensure to copy the existing version into new directory structure for first commit (including refactoring) and then push new changes including version updates in separate commits. This is required to review the changes efficiently.
  - Adding a new service

- [x] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by roslynlu** [comment]
Hi, I'm creating a new swagger for a new API version, and was wondering if I need to check in the previous version then create a new PR with the changes in order to see the diff? Or will you be able to see the diff if I just add the changes as commits to this PR

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json
Comment (mentat9) [inline specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json:694]: >password [](http://example.com/codeflow?start=9&length=8)

Tag this with x-ms-secret extension

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json
Comment (mentat9) [inline specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json:823]: >additionalProperties [](http://example.com/codeflow?start=11&length=20)

Highly recommended to specify the property names. These properties are hard to use by consumers of the API. They also will not be discoverable by Azure Policy for support in policy definitions.

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json
Comment (mentat9) [inline specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json:866]: >additionalProperties [](http://example.com/codeflow?start=11&length=20)

Same. Use real properties if possible.

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/inputs.json
Comment (mentat9) [inline specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/inputs.json:956]: >Identity [](http://example.com/codeflow?start=5&length=8)

This definition appears to be unreferenced.

Comment (mentat9) [reply specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/inputs.json:956]: Actually, I just found the reference over in streamingJobs.json. Might want to consider moving it over there, since it's referenced there and not in this file.

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/operations.json
Comment (mentat9) [inline specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/operations.json:53]: Needs a default response. [https://armwiki.azurewebsites.net/rp_onboarding/process/api_review_best_practices.html#common-issues-found-in-review](https://armwiki.azurewebsites.net/rp_onboarding/process/api_review_best_practices.html#common-issues-found-in-review)

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/streamingjobs.json
Comment (mentat9) [inline specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/streamingjobs.json:477]: >responses [](http://example.com/codeflow?start=9&length=9)

Needs a default response.

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json
Comment (roslynlu) [reply specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json:694]: Hi, can you link me to the documentation for this extension? I don't see it here, https://github.com/Azure/autorest/blob/master/docs/extensions/readme.md, is this doc out of date?

**Comment by roslynlu** [comment]
@mentat9 
Thank you for your feedback! 2 Questions: 
1. How do I handle the cross version breaking change errors? For example, we did not have the default response code for our APIs before, so I added them, but its flagging it as a breaking change. Or if I updated the name of a definition?
2. For some reason, I cannot add labels to my PR so I am unable to add the WaitForARMFeedback label. How can I get this permission?

**Comment by mentat9** [comment]
> @mentat9
> Thank you for your feedback! 2 Questions:
> 
> 1. How do I handle the cross version breaking change errors? For example, we did not have the default response code for our APIs before, so I added them, but its flagging it as a breaking change. Or if I updated the name of a definition? Or when it compares to our preview version for example, we have features that are in our preview version that we do not yet want to include in this GA version, so these differences are getting flagged as well.
> 2. For some reason, I cannot add labels to my PR so I am unable to add the WaitForARMFeedback label. How can I get this permission?

1. For SDK breaking changes, the SDK reviewers will help you with that. If it's a REST breaking change, you would need to get a breaking change review, but from what I saw previously, I don't think you had any REST breaking changes.
2. This is unexpected and makes it hard for me know know when to resume. I'll circle back since I saw your comment.

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/inputs.json
Comment (mentat9) [inline specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/inputs.json:1065]: Removing CustomClr from this enum is a new breaking change. Was it covered by your breaking change review/approval?

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/functions.json
Comment (mentat9) [inline specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/functions.json:488]: >required [](http://example.com/codeflow?start=7&length=8)

Changing a property to required is a new breaking change. Was it covered by your breaking change review/approval?

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/functions.json
Comment (mentat9) [inline specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/functions.json:559]: >required [](http://example.com/codeflow?start=7&length=8)

Changing a property to required is a new breaking change. Was it covered by your breaking change review/approval?

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/functions.json
Comment (mentat9) [inline specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/functions.json:689]: >required [](http://example.com/codeflow?start=7&length=8)

Changing a property to required is a new breaking change. Was it covered by your breaking change review/approval?

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json
Comment (mentat9) [reply specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json:730]: Yes, x-ms-secret should be documented there. Anyway, it's just a boolean flag used to mark secrets:
"x-ms-secret": true. I created this issue to get the documentation updated: [https://github.com/Azure/autorest/issues/4115](https://github.com/Azure/autorest/issues/4115)

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json
Comment (mentat9) [inline specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json:730]: >password [](http://example.com/codeflow?start=9&length=8)

There is a bigger issue with this property. Secrets are not allowed in PUT, PATCH or GET responses, but this type is part of the "Output" definition that is present in all of those payloads. This makes it doubly important you tag it with the x-ms-secret extension.

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json
Comment (mentat9) [inline specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json:780]: It looks like the Microsoft.AzureFunction discriminator schema has been lost in moving forward from 2017-04-01: "x-ms-discriminator-value": "Microsoft.AzureFunction". That would be another breaking change. Was this intentional?

**Comment by mentat9** [comment]
> > @mentat9
> > Thank you for your feedback! 2 Questions:
> > 
> > 1. How do I handle the cross version breaking change errors? For example, we did not have the default response code for our APIs before, so I added them, but its flagging it as a breaking change. Or if I updated the name of a definition? Or when it compares to our preview version for example, we have features that are in our preview version that we do not yet want to include in this GA version, so these differences are getting flagged as well.
> > 2. For some reason, I cannot add labels to my PR so I am unable to add the WaitForARMFeedback label. How can I get this permission?
> 
> 1. For SDK breaking changes, the SDK reviewers will help you with that. If it's a REST breaking change, you would need to get a breaking change review, but from what I saw previously, I don't think you had any REST breaking changes.
> 2. This is unexpected and makes it hard for me know know when to resume. I'll circle back since I saw your comment.

Your latest changes now do include several REST breaking changes where I've commented. Please indicate whether those are intentional: I'll apply the correct labels, and you will need to get a breaking change approval. Otherwise, revert the breaking changes and address the other comments to proceed.

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/functions.json
Comment (roslynlu) [reply specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/functions.json:488]: I changed these to required because I was getting a validation error that said all discriminator values must be required.

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json
Comment (roslynlu) [reply specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json:730]: We made the decision a long time ago to only have one definition for all the requests to make the SDK experience easier for customers. Our workaround was to add more information in the description such as "Required on PUT (CreateOrReplace) requests" to indicate that this property is required on PUT requests, but not PATCH requests, for example. Can we keep it like this to maintain consistency?

Comment (mentat9) [reply specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json:730]: Removing it from the response would break existing clients, so I won't block approval on it. Your RP should at least blank or mask it from GET, PUT, and PATCH responses. If there's a need to retrieve it, you should only return it in the POST action.

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json
Comment (roslynlu) [reply specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json:818]: @mentat9 
Is there a way to enforce this in the swagger? Or do you just mean specify them in the definition? For example, for this property systemPropertyColumns, the object will look something like this:
            "systemPropertyColumns": { 
              "MessageId": "col3", 
              "PartitionKey": "col4"
            }
Where the key can only be certain system properties that we support. I don't think I'm aware of any ways to specify/constrain the possible keys in an object.

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/functions.json
Comment (roslynlu) [reply specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/functions.json:488]: @mentat9 I changed these properties to required because the validation tool was throwing an error saying that discriminator values must be required -- is this not the case? If not, I can change it back

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json
Comment (roslynlu) [reply specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json:730]: Yes, we made the conscious decision to only use one definition for all of our apis to make it more intuitive for customers. So in this case, we only have one Output definition, instead of having a PatchOutput definition, and another for GetOutput, etc. I would prefer to leave it like this to maintain consistency.

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json
Comment (roslynlu) [reply specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json:780]: Yes, this was intentional. We do not want to support Azure Functions in this version.

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json
Comment (mentat9) [reply specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json:818]: No, I mean specify them in the object definition. Even specifying some of them is better than none.

ARM recommends against the use of additionalProperties. Open-ended unschematized objects are problematic, even though they appear to make some things easier for RP owners. ARM is evolving toward stronger typing wherever possible because it allows all consumers of your API to reason about it. This has benefits for everyone that interacts with your API from infrastructure, to other Azure features to SDK owners to customers. Many APIs use this, but it's a pattern that should be avoided if at all possible.

**Comment by mentat9** [comment]
Please work with this week's oncall @pilor to complete the review.

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/inputs.json
Comment (pilor) [inline specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/inputs.json:1021]: why doesn't this take a resource ID to the server/database?  How will it know how to connect to a server based on a name?

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/inputs.json
Comment (pilor) [inline specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/inputs.json:1033]: please mark as a secret and ensure it isn't returned on responses or GETs

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json
Comment (pilor) [inline specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json:491]: timeWindow and sizeWindow need descriptions

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json
Comment (pilor) [inline specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json:714]: same comment as the SQLReference data source.  Why not a resource ID and how do you know how to connect purely based on a server name and database?

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json
Comment (pilor) [inline specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json:1010]: "writer"

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/inputs.json
Comment (pilor) [inline specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/inputs.json:706]: this needs to be marked as a secret and ensure it doesn't get returned in responses

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json
Comment (atpham256) [reply specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json:730]: If I understand the original comment, the concern is that secrets are returned the **response** payload of PUT, PATCH, or GET responses. I can guarantee we are following that. Although it may be supplied as part of the request payload, we remove secret values from the response before sending it back to the customer (which has garnered a number of customer complaints for us btw :) although I understand the security aspect of the requirement)

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json
Comment (atpham256) [reply specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/outputs.json:780]: I think it would be more appropriate to review from the perspective of the previous stable API version as that is what customers are expected to upgrade from.

There are things in preview that we may have introduced that we decided to not move forward with moving to stable.

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/inputs.json
Comment (atpham256) [reply specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/inputs.json:1033]: In general, we are following the rule that no secrets are returned in any response payloads. We currently do not have any APIs that will return these secrets once you set them.

File: specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/inputs.json
Comment (roslynlu) [reply specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/inputs.json:1021]: To my knowledge, any user who wants to connect to a sql server only needs to provide these properties (servername, dbname, username, password). We are just using their SDKs: https://docs.microsoft.com/en-us/dotnet/api/overview/azure/sql

Comment (pilor) [reply specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2020-03-01/inputs.json:1021]: It may work but its going to require the user to pass around secrets which is generally discouraged.  I won't block on this but I suggest you bring this up in your next thread modeling/security review.  Ideally this uses a managed identity that is attached to your resource to access the SQL server via AAD auth (https://docs.microsoft.com/en-us/azure/azure-sql/database/authentication-aad-overview)

**Comment by zesluo** [comment]
Does the cluster and non-cluster both use  "2020-03-01" as their version ? Thanks

**Review by JeffreyRichter — APPROVED** [review]

**Comment by zhenglaizhang** [comment]
@roslynlu pls fix the LintDiff errors in https://github.com/Azure/azure-rest-api-specs/pull/14199#issuecomment-841421371

**Comment by msyyc** [comment]
For python, the breaking change is expected.

**Pull request merged**
